src/lib/db/public-signup-token-store.ts

```diff
@@ -0,0 +1,193 @@
+            users: [],
+        };
+    }
+    const currentToken = acc[tokenRow.secret];
+    if (userId) {
+        currentToken.users.push({ userId, name });
+    }
+    return acc;
+};
+
+const toRow = (newToken: IPublicSignupTokenCreate) => ({
+    secret: newToken.secret,
+    name: newToken.name,
+    expires_at: newToken.expiresAt,
+    created_by: newToken.createdBy || null,
+    role_id: newToken.roleId,
+});
+
+const toTokens = (rows: any[]): PublicSignupTokenSchema[] => {
+    const tokens = rows.reduce(tokenRowReducer, {});
+    return Object.values(tokens);
+};
+
+export class PublicSignupTokenStore implements IPublicSignupTokenStore {
+    private logger: Logger;
+
+    private timer: Function;
+
+    private db: Knex;
+
+    constructor(db: Knex, eventBus: EventEmitter, getLogger: LogProvider) {
+        this.db = db;
+        this.logger = getLogger('public-signup-tokens.js');
+        this.timer = (action: string) =>
+            metricsHelper.wrapTimer(eventBus, DB_TIME, {
+                store: 'public-signup-tokens',
+                action,
+            });
+    }
+
+    count(): Promise<number> {
+        return this.db(TABLE)
+            .count('*')
+            .then((res) => Number(res[0].count));
+    }
+
+    private makeTokenUsersQuery() {
+        return this.db<ITokenRow>(`${TABLE} as tokens`)
+            .leftJoin(
+                `${TOKEN_USERS_TABLE} as token_project_users`,
+                'tokens.secret',
+                'token_project_users.secret',
+            )
+            .leftJoin(`users`, 'token_project_users.user_id', 'users.id')
+            .select(
+                'tokens.secret',
+                'tokens.name',
+                'tokens.expires_at',
+                'tokens.created_at',
+                'tokens.created_by',
+                'tokens.role_id',
+                'token_project_users.user_id',
+                'users.name',
+            );
+    }
+
+    async getAll(): Promise<PublicSignupTokenSchema[]> {
+        const stopTimer = this.timer('getAll');
+        const rows = await this.makeTokenUsersQuery();
+        stopTimer();
+        return toTokens(rows);
+    }
+
+    async getAllActive(): Promise<PublicSignupTokenSchema[]> {
+        const stopTimer = this.timer('getAllActive');
+        const rows = await this.makeTokenUsersQuery()
+            .where('expires_at', 'IS', null)
+            .orWhere('expires_at', '>', 'now()');
+        stopTimer();
+        return toTokens(rows);
+    }
+
+    async addTokenUser(secret: string, userId: number): Promise<void> {
+        await this.db.transaction(async (tx) => {
+            await tx<ITokenUserRow>(TOKEN_USERS_TABLE).insert(
+                { user_id: userId, secret },
+                ['created_at'],
+            );
+        });
+    }
+
+    async insert(
+        newToken: IPublicSignupTokenCreate,
+    ): Promise<PublicSignupTokenSchema> {
+        const response = await this.db.transaction(async (tx) => {
```

Comment:
I still think we should remove this transaction, it only adds overhead when there is only one call to the database.